add space reset in openvr bindings

meir opened the pull request:
Tried to add the spacedrag bindings in SteamVR yesterday but found out the option for the space reset is missing.

This should add it in SteamVR.

Will have to test when i get back home, ill keep the PR in draft until ive managed to test it.

meir commented:
Tested, works in SteamVR now, however there is a bug where the DragReset puts you under the floor and away from the center for some reason

galister commented:
If you want to look further into this, here's some info:

Moving the playspace on SteamVR consists of moving the actual chaperone (the boundary thing that your draw during steamvr room calibration).

When uncalibrated, the chaperone sits at the raw center of the playspace. (This is likely what it resets to as of now.)

Upon room calibration, the chaperone is saved to a file somewhere, and SteamVR will use this calibration for the chaperone on next launch.

Instead of resetting the chaperone by clearing its transform, we want to revert it to the pose that's in the calibration file.

There's an OpenVR method that seems to do just what we need: [IVRChaperoneSetup::ReloadFromDisk](https://github.com/ValveSoftware/openvr/blob/ae46a8dd0172580648c8922658a100439115d3eb/headers/openvr.h#L3315)

meir commented:
ahh alright, i'll see tomorrow if i can figure it out, might need to check my actual boundary in SteamVR aswell, i might have it improperly set since OVRAS sometimes broke it and i got sick of fixing it each time it happened

meir commented:
couldn't really stop thinking about it, especially since ive recently started learning rust so got a bit too excited about it

but looked into it and ReloadFromDisk wouldnt work since the CommitWorkingCopy would override the saved chaperone, this would technically mean that its overridden entirely, so i changed it out for the working preview so that the saved chaperone stays in tact and does not get ruined

However this does come with another bug, it's only visual and not breaking, but when the preview is showing, the chaperone bounds will just fly away, i can't really see why yet, but the play space stays at its origin

I'll look later into why this might be happening

meir commented:
found the issue with the boundaries flying off, adding an offset to them was breaking it, apparently steamvr already sets the offset when previewing the working set.

galister commented:
i'm not sure what the nix part is about, we're already included in: https://github.com/nix-community/nixpkgs-xr

if it's needed then please make it into its own PR

meir commented:
i added the nix flake for local development, the flake in nixpkgs-xr builds a specific git branch, the one i added is to run/build the local code using nix

meir commented:
it is probably better to override the wlx package just like nixpkgs-xr does, im not sure if theres another option in nix to do local development using the nixpkgs-xr

meir committed:
fix openvr spacereset and add openvr bindings for spacereset

meir commented:
found another issue before this pr gets merged, when moving, any windows shift with you, when resetting they dont, i'll figure this out later today

meir committed:
fix spacedrag by adding offset value to overlaydata

meir commented:
changed how the space drag moves the overlays, added an offset value to the overlay data instead which can in turn be easily reset instead.

Tested in SteamVR, works exactly as expected

meir committed:
Merge branch 'galister:main' into main

meir committed:
fix cursor not hitting when playspace is dragged in OpenVR

meir committed:
Merge branch 'galister:main' into main

meir committed:
Merge branch 'main' into space-reset-ovr

meir committed:
Merge branch 'galister:main' into space-reset-ovr

meir committed:
Merge branch 'galister:main' into space-reset-ovr

meir committed:
Merge branch 'galister:main' into space-reset-ovr

meir committed:
Merge pull request #1 from meir/space-reset-ovr

Space reset ovr

galister commented:
Thanks, it's starting to look pretty good.

Some ideas from me:

Let's keep `OverlayState`'s `transform` as the effective transform. We don't want to calculate an effective transform every time.

We should store the inverse offset so that from the effective transform, we can calculate the original transform (and offset) when needed.

Since these offsets are only needed for SteamVR, let's store them in OpenVrOverlayData. (`overlay.data` instead of `overlay.state`)

meir committed:
Merge branch 'galister:main' into space-reset-ovr

meir committed:
change offset to be in OVR data and just to keep track of the offset instead of applying it to the translation

meir committed:
Merge pull request #2 from meir/space-reset-ovr

Space reset ovr

meir committed:
change offset to be in OVR data and just to keep track of the offset instead of applying it to the translation

meir committed:
Merge branch 'staging' into space-reset-ovr

meir committed:
use overlay.data.offset

meir commented:
tested it out yesterday, everything seems to work like a dream

meir committed:
Merge branch 'galister:main' into space-reset-ovr

galister commented on src/backend/openvr/lines.rs:
```diff
@@ -83,6 +83,7 @@ impl LinePool {
                 override_width: true,
                 image_view: Some(self.view.clone()),
                 image_dirty: true,
+                offset: Affine3A::IDENTITY,
```
Is this needed? Lines are only ever used for pointer lasers, and their transforms are recalculated every frame.

meir replied:
no it is not, not sure why i added this, i'll fix this after testing the space rotate tonight so i can keep it in the same commit

galister commented on src/backend/openvr/playspace.rs:
```diff
@@ -73,17 +72,15 @@ impl PlayspaceMover {
                     overlay.state.dirty = true;
                     overlay.state.transform.translation =
                         overlay_transform.transform_point3a(overlay.state.transform.translation);
+                    overlay.data.offset.translation =
+                        overlay_transform.transform_point3a(overlay.state.transform.translation);
```
Have you had a chance to test the behavior of this with a `space_rotate`'d playspace?

I would wager you might run into issues when rotations are involved.

The equivalent matrix operation would be something along the lines of:
`overlay.data.offset = overlay_transform * overlay.state.transform`

meir replied:
ah, i completely forgot about this, i never really use space rotate, i'll test this out tonight

galister replied:
One of the things about space rotate is that the space actually rotates around the center of the playspace, but we make it seem as it's rotating around the HMD. This is done by counter-adjusting the offset position to keep the HMD in the same position. I believe this will mess stuff up. At the very least, you'd need to also write the offset from the space_rotate logic.

meir replied:
tested this out yesterday, 2 things to note

1. the reset is indeed broken, so ill update this with some code to account for rotation
2. the rotation of the overlays seems to be broken aswell, they do move, but dont rotate with the space rotate, so ill look into this aswell

meir committed:
Merge branch 'main' of https://github.com/meir/wlx-overlay-s into space-reset-ovr

meir committed:
Merge branch 'galister:main' into space-reset-ovr
